### PR TITLE
Update T5 Onnx Export and Optimization

### PR DESCRIPTION
**Pull request by tianleiwu**
### Description

Previously, the encoder onnx model adds extra initialization for decoder to generate kv cache from prompt. It is not necessary. Here we redesign onnx export for T5 model to output two separate models for encode and decoder.

Move Linear that generates cross features based on encoder_hidden_states to encoder onnx model. In this way, the encoder does not need output encoder_hidden_states, and only need output the features for cross attention used in decoder.

Major changes:
 -[x] update t5 onnx export script
 -[x] update convert_generation script
 -[x] update beam search to support changes of inputs and outputs (detail can be found below).
 -[x] add a tiny t5 model, and enable the generation test for T5 in Linux CI pipelines.

Example change in inputs and outputs for one layer model:
**Encoder Inputs**:
- encoder_input_ids: int32 (B, encode_sequence_length)
- encoder_attention_mask: int32 (B, encode_sequence_length)
- ~~decoder_input_ids: int32 (B, 1)~~

**Encoder Outputs**:
- ~~logits: (B, 1, vocab_size)~~
- ~~encoder_hidden_states: (B, encode_sequence_length, encoder_hidden_size)~~
- ~~present_key_self_0: (B, num_heads, 1, head_size)~~
- ~~present_value_self_0: (B, num_heads, 1, head_size)~~
- present_key_cross_0: (B, num_heads, encode_sequence_length, head_size)
- present_value_cross_0: (B, num_heads, encode_sequence_length, head_size)

**Decoder Inputs**:
- input_ids: int32 (B, 1)
- ~~encoder_input_ids: int32 (B, encode_sequence_length) (optional for old format; removed in new format)~~
- encoder_attention_mask: int32 (B, encode_sequence_length)
- ~~encoder_hidden_states: (B, encode_sequence_length, encoder_hidden_size) (optional for old format; removed in new format)~~
- past_key_self_0: (B, num_heads, past_decode_sequence_length, head_size)
- past_value_self_0: (B, num_heads, past_decode_sequence_length, head_size)
- past_key_cross_0: (B, num_heads, encode_sequence_length, head_size)
- past_value_cross_0: (B, num_heads, encode_sequence_length, head_size)

**Decoder Outputs**:
- logits: (B, 1, vocab_size)
- present_key_self_0: (B, num_heads, past_decode_sequence_length + 1, head_size)
- present_value_self_0: (B, num_heads, past_decode_sequence_length + 1, head_size)

Known issues:
- Some postprocessing (like converting to use decoder masked MHA, past and present buffer sharing) is not done. Could be a future work item to integrate with onnxruntime-genai.

### Motivation and Context

Make the encoder onnx model simpler and more efficient in inference (no need to output encoder_hidden_states).

### CODE REVIEW

File: onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc [hunk 1]
```diff
@@ -141,14 +141,23 @@ Status T5DecoderSubgraph::Validate(const std::vector<const NodeArg*>& subgraph_i
...
-//                encoder_attention_mask,
-//                encoder_hidden_states,
-//                present_key_self_0, present_value_self_0, ..., present_key_cross_0, present_value_cross_0, ...
+// New format:
+//   encoder feeds: encoder_input_ids, encoder_attention_mask
+//   encoder fetches: present_key_cross_0, present_value_cross_0, ...
+//   decoder_feeds: input_ids, encoder_attention_mask,
+//                  present_key_self_0, present_value_self_0, ...,
+//                  present_key_cross_0, present_value_cross_0, ...
+
```
Comment (kunal-vaishnavi) [inline onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc:150]: Can we add what the decoder fetches are since we list the encoder's feeds and fetches?

File: onnxruntime/python/tools/transformers/models/t5/convert_to_onnx.py [hunk 2]
```diff
@@ -193,6 +214,16 @@ def export_onnx_models(
 
         # Optimize ONNX graph. Note that we have not implemented graph optimization for T5 yet.
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/models/t5/convert_to_onnx.py:215]:
```suggestion
        # Optimize ONNX graph.
```

File: onnxruntime/python/tools/transformers/models/t5/t5_encoder_decoder_init.py [hunk 3]
```diff
@@ -108,9 +116,14 @@ def export_onnx(
             onnx_model_path (str): onnx path
             verbose (bool, optional): print verbose information. Defaults to True.
             use_external_data_format (bool, optional): use external data format or not. Defaults to False.
+            use_int32_inputs (bool, optional): use int32 instead of int64 for integer inputs. Defaults to False.
         """
         assert isinstance(model, T5EncoderDecoderInit)
 
+        # Do not exclude decoder in torch onnx export so that cross can show up.
+        output_cross_only = model.output_cross_only
+        model.output_cross_only = False
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/models/t5/t5_encoder_decoder_init.py:125]: If these methods can be used for other models, can we add the methods in [`past_helper.py`](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/python/tools/transformers/past_helper.py) instead for other models to leverage them?

**Comment by kunal-vaishnavi** [comment]
Adding a note here to check that the changes to the encoder and decoder subgraphs for T5 do not affect Whisper. There are CI tests [here](https://github.com/microsoft/onnxruntime/blob/dcc1f5ac59ca292267871517f6442648a7dbe120/tools/ci_build/github/azure-pipelines/bigmodels-ci-pipeline.yml#L435) that were added to ensure backwards compatibility with the `WhisperBeamSearch` op.

File: onnxruntime/python/tools/transformers/onnx_model_t5.py [hunk 4]
```diff
@@ -268,23 +278,32 @@ def fuse_t5_encoder(self, normalize_node, input_name_to_nodes, output_name_to_no
...
             if mask_nodes is None:
-                return
+                mask_nodes = self.model.match_parent_path(
+                    add_qk,
+                    ["Add", "Slice", "Mul", "Sub", "Cast", "Unsqueeze", "Unsqueeze"],
+                    [1, 1, 0, 0, 1, 0, 0],
+                    output_name_to_node,
+                )
+                if mask_nodes is None:
+                    return
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/onnx_model_t5.py:301]: Why mix return types? Some cases use `return False` while other cases use `return`

File: onnxruntime/python/tools/transformers/onnx_model_t5.py [hunk 5]
```diff
@@ -712,9 +758,50 @@ def fuse(self, node, input_name_to_nodes, output_name_to_node):
...
+        #
+        #      total_seq_len    Range(start=past_seq_len, end=total_seq_len)
+        #              |              |
+        #          Range(0, *)   Unsqueeze(axes=1)
+        #              |              |
+        #    Unsqueeze(axes=0)    Cast(to=int64)
+        #                   \     /
+        #                     Sub
+        #
+        # Founatutionally, there is still Slice to get last seq_len rows so end result is same.
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/onnx_model_t5.py:794]:
```suggestion
        # Foundationally, there is still Slice to get last seq_len rows so end result is same.
```

File: onnxruntime/test/python/transformers/test_generation.py [hunk 6]
```diff
@@ -233,55 +253,163 @@ def remove_onnx_files(self):
...
+    def check_decoder_fusion(self):
+        model_name = self.model_name
+        onnx_path = os.path.join(".", "onnx_models", f"{model_name}_decoder_fp16.onnx")
+
+        model = onnx.load_model(onnx_path, format=None, load_external_data=True)
+        from onnxruntime.transformers.onnx_model import OnnxModel
+
+        onnx_model = OnnxModel(model)
+        op_counters = onnx_model.get_operator_statistics()
+        print("decoder opators", op_counters)
```
Comment (kunal-vaishnavi) [inline onnxruntime/test/python/transformers/test_generation.py:354]:
```suggestion
        print("decoder ops", op_counters)
```

**Comment by kunal-vaishnavi** [comment]
Can we add T5 to the `Big Models` CI Pipeline?

**Comment by tianleiwu** [comment]
> Can we add T5 to the `Big Models` CI Pipeline?
It is not needed. The T5 tests in test_generation.py is enabled in CI pipeline so there is end to end tests there.

File: onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc (hunk 1 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc]: This is a function to create feeds. The fetches (outputs) can be found at the top of this file.

File: onnxruntime/python/tools/transformers/models/t5/t5_encoder_decoder_init.py (hunk 3 above)
Comment (tianleiwu) [reply onnxruntime/python/tools/transformers/models/t5/t5_encoder_decoder_init.py:125]: We can do some refactoring when we apply new format for other models (like bart or whisper). Maybe create a base class for encoder decoder export.

File: onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc [hunk 7]
```diff
@@ -173,47 +182,50 @@ Status T5DecoderSubgraph::CreateInitialFeeds(
...
             sequence,
             stream,
             DeviceCopyDirection::deviceToDevice));
       }
     } else {
-      const size_t cur_len_bytes = cur_len * sizeof(int);
+      size_t total_size = static_cast<size_t>(sequence_length) * static_cast<size_t>(batch_beam_size);
+      size_t total_size_bytes = total_size * sizeof(int);
+      AllocatorPtr buffer_allocator = std::make_shared<onnxruntime::CPUAllocator>();
+      // TODO: not need extra buffer. Copy directly to input_ids_data instead like the user_cuda above.
```
Comment (kunal-vaishnavi) [inline onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc:219]: In another PR or this PR?

File: onnxruntime/python/tools/transformers/convert_generation.py [hunk 8]
```diff
@@ -223,6 +237,14 @@ def parse_arguments(argv: list[str] | None = None) -> argparse.Namespace:
     )
     output_group.set_defaults(disable_shared_initializers=False)
 
+    output_group.add_argument(
+        "--encode_decoder_init",
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/convert_generation.py:241]:
```suggestion
        "--encoder_decoder_init",
```

File: onnxruntime/python/tools/transformers/models/t5/convert_to_onnx.py [hunk 9]
```diff
@@ -131,34 +138,52 @@ def parse_arguments():
         help="filepath to load pre-trained model with custom state dictionary (e.g. pytorch_model.bin)",
     )
 
+    parser.add_argument(
+        "--encode_decoder_init",
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/models/t5/convert_to_onnx.py:142]:
```suggestion
        "--encoder_decoder_init",
```

File: onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc (hunk 7 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc:219]: In another PR.

**Review by kunal-vaishnavi — APPROVED** [review]
